### PR TITLE
WIP: Nms/Obc work

### PR DESCRIPTION
**Pull request by pisaiah**
Classes are for 1.12.2, I'll update them to 1.13 when Glowstone & Spigot 1.13 are stable

Loads the glowstone.jar using Class loader
Place [linkstone-1.0.0.jar](https://ci.appveyor.com/project/IsaiahPatton/linkstone/build/artifacts) next to your glowstone.jar and run linkstone.jar instead of glowstone.

[Build status](https://ci.appveyor.com/project/IsaiahPatton/linkstone)

Bukkit.getServer() will now return an instance of CraftServer instead of GlowServer.
(_Bukkit.getNonLinkstone() is used to get GlowServer_), This allows casting Server to CraftServer, Player to CraftPlayer, Entity to CraftEntity, etc ...

## Out of the classes I've done so far:

**Full working NMS/OBC classes:**

- org/bukkit/Bukkit.java **(for casting)**
- net/glowstone/ServerProvider.java **(Had to fix crash, with getServer being CraftServer now)**
- org/bukkit/command/PluginCommand.java **(allow casting CommandSender to CraftPlayer)**
- com/mojang/authlib/GameProfile.java
- com/mojang/authlib/properties/PropertyMap.java
- net/minecraft/server/EnumDifficulty.java
- net/minecraft/server/IAsyncTaskHandler.java
- net/minecraft/server/IAnimal.java
- net/minecraft/server/IBlockAccess.java
- net/minecraft/server/IMojangStatistics.java
- net/minecraft/server/IMonster.java
- net/minecraft/server/IWorldBorderListener.java
- net/minecraft/server/ReportedException.java
- net/minecraft/server/Vec3D.java
- org/bukkit/craftbukkit/CraftServer.java
- org/bukkit/craftbukkit/CraftWorld.java
- org/bukkit/craftbukkit/Main.java
- org/bukkit/craftbukkit/Overridden.java
- org/bukkit/craftbukkit/boss/CraftBossBar.java
- org/bukkit/craftbukkit/entity/CraftEntity.java
- org/bukkit/craftbukkit/entity/CraftHumanEntity.java
- org/bukkit/craftbukkit/entity/CraftLivingEntity.java
- org/bukkit/craftbukkit/entity/CraftPlayer.java
- org/bukkit/craftbukkit/entity/CraftVehicle.java
- org/bukkit/craftbukkit/help/CustomHelpTopic.java
- org/bukkit/craftbukkit/util/CraftMagicNumbers.java
- org/bukkit/craftbukkit/util/CraftNamespacedKey.java
- org/bukkit/craftbukkit/util/DatFileFilter.java
- org/bukkit/craftbukkit/util/permissions/CommandPermissions.java
- org/bukkit/craftbukkit/util/permissions/CraftDefaultPermissions.java

**Still needs work**

- net/minecraft/server/MathHelper.java
- net/minecraft/server/MinecraftKey.java
- net/minecraft/server/MinecraftServer.java
- net/minecraft/server/MojangStatisticsGenerator.java (**should we send data to Mojang?**)
- net/minecraft/server/Registry.java
- net/minecraft/server/RegistryID.java
- net/minecraft/server/World.java
- net/minecraft/server/RegistryMaterials.java
- net/minecraft/server/RegistrySimple.java
- net/minecraft/server/TileEntity.java
- net/minecraft/server/WorldBorder.java
- net/minecraft/server/WorldServer.java
- org/apache/logging/log4j/LogManager.java
- org/apache/logging/log4j/Logger.java
- org/bukkit/craftbukkit/util/MojangNameLookup.java
- org/bukkit/craftbukkit/util/TerminalConsoleWriterThread.java
- org/bukkit/craftbukkit/util/Versioning.java

### CODE REVIEW

**Comment by kamcio96** [comment]
I suggest edit ASM code to call and cast glowstone class instead craftbukkit

File: src/main/java/net/glowstone/linkstone/Linkstone.java [hunk 1]
```diff
@@ -1,16 +1,71 @@
...
+            if (s.startsWith("glowstone:")) glowstone = s.substring("glowstone:".length()).trim();
+        }
+
+    	File glowjar = new File(glowstone);
+    	addURL(glowjar.toURI().toURL());
+    }
+
+    private static void addURL(URL u) throws IOException {
+        System.out.println("Adding glowstone jar to classpath ...");
+        URLClassLoader sysloader = (URLClassLoader) ClassLoader.getSystemClassLoader();
```
Comment (Cybermaxke) [inline src/main/java/net/glowstone/linkstone/Linkstone.java:46]: The SystemClassLoader is no longer a URLClassLoader in java 9+. So this breaks compatibility with those versions.

Comment (aki-ks) [reply src/main/java/net/glowstone/linkstone/Linkstone.java:46]: Once we migrated the code to my linkstone framework, this code will be dropped.

**Comment by pisaiah** [comment]
#6

**Comment by Pr0methean** [comment]
Re MojangStatisticsGenerator: this should be configurable. Default could be to contact Mojang if and only if we're not in offline mode (since without offline mode, we contact Mojang to authenticate players).

**Pull request closed without merging**